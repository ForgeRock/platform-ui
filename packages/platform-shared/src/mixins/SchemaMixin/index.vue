<!-- Copyright 2020 ForgeRock AS. All Rights Reserved

Use of this code requires a commercial software license with ForgeRock AS.
or with one of its affiliates. All use shall be exclusively subject
to such license between the licensee and ForgeRock AS. -->
<script>
import {
  mapKeys,
  omit,
  has,
} from 'lodash';
/**
 * @description Schema mixin used for filtering and mapping json schema information to be consumed by schema driven components such as FormGenerator.
 */
export default {
  name: 'SchemaMixin',
  methods: {
    /**
     * Takes current schema object and calculates all render types and collects all options for populating arrays
     * @param schema - current schema defined by AM
     */
    filterAndMapRawSchema(schema) {
      const newSchema = {};

      // The useful schema information is always stored under the 'properties' at the top level
      mapKeys(schema.properties, (propertyValue, propertyKey) => {
        newSchema[propertyKey] = this.filterAndMapSchemaProperty(propertyValue);
      });

      return newSchema;
    },
    /**
     * Determines if a schema property is nested
     * @param schemaProperty - the schema property to evaluate
     */
    isNestedProperty(schemaProperty) {
      if (!schemaProperty.properties) {
        return false;
      }

      const propertyKeys = Object.keys(schemaProperty.properties);
      if (propertyKeys.length === 2 && propertyKeys.includes('value') && propertyKeys.includes('inherited')) {
        return false;
      }
      return true;
    },
    /**
     * Recursively filters and maps a potentially nested json schema representation
     * of a field onto something that can be consumed by the FormGenerator component.
     * @param schemaProperty - the property to evaluate
     */
    filterAndMapSchemaProperty(schemaProperty) {
      // If the property is nested, recursively map sub properties
      if (this.isNestedProperty(schemaProperty)) {
        const newObj = {};
        mapKeys(schemaProperty.properties, (prop, name) => {
          newObj[name] = this.filterAndMapSchemaProperty(prop);

          if (newObj[name].required) {
            if (newObj.required && newObj.required.length) {
              newObj.required.push(name);
            } else {
              newObj.required = [name];
            }
          }
        });
        return newObj;
      }

      // if the property is not nested, map it directly
      return this.filterAndMapBaseLevelSchemaProperty(schemaProperty);
    },
    /**
     * Filters and maps a non-nested json schema field onto something that can be
     * consumed by the FormGenerator component.
     * @param prop - the property to evaluate
     */
    filterAndMapBaseLevelSchemaProperty(prop) {
      const mappedProp = omit(prop, ['propertyOrder', 'type', 'exampleValue', 'properties', 'options', 'inherited']);
      mappedProp.type = prop.properties && prop.properties.value && prop.properties.value.type ? prop.properties.value.type : '';

      const hasType = has(mappedProp, 'type');
      const hasEnumNames = has(mappedProp, 'enumNames');
      const hasItems = has(mappedProp, 'items');
      const hasEnumNamesInItems = hasItems && has(mappedProp.items, 'enumNames');
      const enumNamesLengthIs2 = hasEnumNames && mappedProp.enumNames.length === 2;

      if (!hasType) {
        mappedProp.type = prop.type || 'string';
      }

      // assign render type  and array subtype to schema object
      if (enumNamesLengthIs2) {
        mappedProp.type = 'radio';
        mappedProp.options = mappedProp.enumNames.map((text, index) => ({ text, value: mappedProp.enum[index] }));
        delete mappedProp.enum;
        delete mappedProp.enumNames;
      } else if (hasEnumNames) {
        mappedProp.arrayType = 'selectOne';
        mappedProp.type = 'array';
      } else if (hasEnumNamesInItems) {
        mappedProp.arrayType = 'selectMany';
        mappedProp.type = 'array';
        mappedProp.enum = mappedProp.items.enum;
        mappedProp.enumNames = mappedProp.items.enumNames;
        delete mappedProp.items;
      } else if (hasItems) {
        mappedProp.arrayType = 'addMany';
        mappedProp.type = 'array';
        mappedProp.options = [];
        delete mappedProp.items;
      }

      if (prop.properties && prop.properties.value && prop.properties.value.required === true) {
        mappedProp.required = true;
      }
      return mappedProp;
    },
  },
};
</script>
