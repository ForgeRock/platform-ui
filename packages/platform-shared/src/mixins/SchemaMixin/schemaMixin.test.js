/**
 * Copyright 2020 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import { mount } from '@vue/test-utils';
import SchemaMixin from './index';

const exampleApplicationSchema = {
  type: 'object',
  properties: {
    advancedOAuth2ClientConfig: {
      type: 'object',
      title: 'Advanced',
      propertyOrder: 1,
      properties: {
        mixUpMitigation: {
          title: 'OAuth 2.0 Mix-Up Mitigation enabled', description: 'blah', propertyOrder: 26300, type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'boolean', required: true } },
        },
        name: {
          title: 'Display name', description: 'blah', propertyOrder: 23500, items: { type: 'string' }, type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'array', required: false } },
        },
        updateAccessToken: {
          title: 'Access Token', description: 'blah', propertyOrder: 25100, type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'string', required: false } },
        },
        requestUris: {
          title: 'Request uris', description: 'blah', propertyOrder: 23700, items: { type: 'string' }, type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'array', required: false } },
        },
        isConsentImplied: {
          title: 'Implied consent', description: 'blah', propertyOrder: 26200, type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'boolean', required: true } },
        },
        tokenEndpointAuthMethod: {
          title: 'Token Endpoint Authentication Method', description: 'blah', propertyOrder: 24000, enum: ['client_secret_post', 'client_secret_basic', 'private_key_jwt', 'tls_client_auth', 'self_signed_tls_client_auth', 'none'], options: { enum_titles: ['client_secret_post', 'client_secret_basic', 'private_key_jwt', 'tls_client_auth', 'self_signed_tls_client_auth', 'none'] }, enumNames: ['client_secret_post', 'client_secret_basic', 'private_key_jwt', 'tls_client_auth', 'self_signed_tls_client_auth', 'none'], type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'string', required: true } },
        },
        subjectType: {
          title: 'Subject Type', description: 'blah', propertyOrder: 24400, enum: ['pairwise', 'public'], options: { enum_titles: ['pairwise', 'public'] }, enumNames: ['pairwise', 'public'], type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'string', required: true } },
        },
      },
    },
  },
};

const expectedMappedApplicationSchema = {
  advancedOAuth2ClientConfig: {
    isConsentImplied: {
      description: 'blah',
      required: true,
      title: 'Implied consent',
      type: 'boolean',
    },
    mixUpMitigation: {
      description: 'blah',
      required: true,
      title: 'OAuth 2.0 Mix-Up Mitigation enabled',
      type: 'boolean',
    },
    name: {
      arrayType: 'addMany',
      description: 'blah',
      options: [],
      title: 'Display name',
      type: 'array',
    },
    requestUris: {
      arrayType: 'addMany',
      description: 'blah',
      options: [],
      title: 'Request uris',
      type: 'array',
    },
    required: [
      'mixUpMitigation',
      'isConsentImplied',
      'tokenEndpointAuthMethod',
      'subjectType',
    ],
    subjectType: {
      description: 'blah',
      options: [
        {
          text: 'pairwise',
          value: 'pairwise',
        },
        {
          text: 'public',
          value: 'public',
        },
      ],
      required: true,
      title: 'Subject Type',
      type: 'radio',
    },
    tokenEndpointAuthMethod: {
      arrayType: 'selectOne',
      description: 'blah',
      enum: [
        'client_secret_post',
        'client_secret_basic',
        'private_key_jwt',
        'tls_client_auth',
        'self_signed_tls_client_auth',
        'none',
      ],
      enumNames: [
        'client_secret_post',
        'client_secret_basic',
        'private_key_jwt',
        'tls_client_auth',
        'self_signed_tls_client_auth',
        'none',
      ],
      required: true,
      title: 'Token Endpoint Authentication Method',
      type: 'array',
    },
    updateAccessToken: {
      description: 'blah',
      title: 'Access Token',
      type: 'string',
    },
  },
};

const exampleGatewaySchema = {
  type: 'object',
  properties: {
    igCdssoRedirectUrls: {
      title: 'Redirect URLs for CDSSO', description: 'blah', propertyOrder: 150, items: { type: 'string' }, type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'array', required: false } },
    },
    igTokenIntrospection: {
      title: 'Token Introspection', description: 'blah', propertyOrder: 160, enum: ['None', 'Realm', 'Realm_Subs'], options: { enum_titles: ['None', 'Realm Only', 'Realm and Sub Realms'] }, enumNames: ['None', 'Realm Only', 'Realm and Sub Realms'], type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'string', required: false } },
    },
    userpassword: {
      title: 'Password', description: 'blah', propertyOrder: 100, required: true, type: 'string', format: 'password', exampleValue: '',
    },
    status: {
      title: 'Status', description: 'blah', propertyOrder: 200, enum: ['Active', 'Inactive'], options: { enum_titles: ['Active', 'Inactive'] }, enumNames: ['Active', 'Inactive'], type: 'object', exampleValue: '', properties: { inherited: { type: 'boolean', required: true }, value: { type: 'string', required: true } },
    },
  },
};

const expectedMappedGatewaySchema = {
  igCdssoRedirectUrls: {
    arrayType: 'addMany',
    description: 'blah',
    options: [],
    title: 'Redirect URLs for CDSSO',
    type: 'array',
  },
  igTokenIntrospection: {
    arrayType: 'selectOne',
    description: 'blah',
    enum: [
      'None',
      'Realm',
      'Realm_Subs',
    ],
    enumNames: [
      'None',
      'Realm Only',
      'Realm and Sub Realms',
    ],
    title: 'Token Introspection',
    type: 'array',
  },
  status: {
    description: 'blah',
    options: [
      {
        text: 'Active',
        value: 'Active',
      },
      {
        text: 'Inactive',
        value: 'Inactive',
      },
    ],
    required: true,
    title: 'Status',
    type: 'radio',
  },
  userpassword: {
    description: 'blah',
    format: 'password',
    required: true,
    title: 'Password',
    type: '',
  },
};

let wrapper;

beforeEach(() => {
  wrapper = mount({
    template: '<div>Testing is awesome!</div>',
    mixins: [SchemaMixin],
  });
});

describe('Schema mixin', () => {
  describe('isNestedProperty method', () => {
    it('identifies properties with more than two sub-properties as nested', () => {
      expect(wrapper.vm.isNestedProperty(
        {
          something: 'something',
          properties: {
            value: 'test',
            inherited: 'test',
            other: 'test',
          },
        },
      )).toBe(true);
    });

    it('identifies properties which have no sub properties as not nested', () => {
      expect(wrapper.vm.isNestedProperty(
        {
          something: 'something',
        },
      )).toBe(false);
    });

    it('identifies properties which only have \'inherited\' and \'value\' as keys as not nested', () => {
      expect(wrapper.vm.isNestedProperty(
        {
          something: 'something',
          properties: {
            inherited: 'test',
            value: 'test',
          },
        },
      )).toBe(false);
    });
  });

  describe('filtering and mapping json schemas', () => {
    it('correctly filters and maps nested schema information', () => {
      expect(wrapper.vm.filterAndMapRawSchema(exampleApplicationSchema)).toEqual(expectedMappedApplicationSchema);
    });

    it('correctly filters and maps non-nested schema information', () => {
      expect(wrapper.vm.filterAndMapRawSchema(exampleGatewaySchema)).toEqual(expectedMappedGatewaySchema);
    });
  });
});
