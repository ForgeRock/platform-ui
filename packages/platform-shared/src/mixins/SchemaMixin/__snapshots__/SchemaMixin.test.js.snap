// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema mixin isNestedProperty method identifies properties with more than two sub-properties as nested: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation CONFIG_WHITESPACE) Vue 3 compiler's whitespace option will default to "condense" instead of "preserve". To suppress this warning, provide an explicit value for \`config.compilerOptions.whitespace\`.",
    "
",
    " at <Anonymous",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
